### PR TITLE
Picklable generators

### PR DESCRIPTION
**Pull request by ferrine**
Now generators are easy to use when you define generator itself. This routine work can be done in a picklable class. Comprehensive documentation is provided.

### CODE REVIEW

File: pymc3/data.py [hunk 1]
```diff
@@ -81,3 +88,61 @@ def __eq__(self, other):
...
+    >>> with pm.Model():
+    ...     sd = pm.Uniform('sd', 0, 10)
+    ...     mu = pm.Normal('mu', sd=10)
+    ...     obs_norm = pm.Normal('obs_norm', mu=mu, sd=sd,
+    ...                          observed=minibatches,
+    ...                          total_size=data.shape[0])
+    ...     adam = partial(pm.adam, learning_rate=.8) # easy problem
+    ...     approx = pm.fit(10000, method='advi', obj_optimizer=adam)
+    >>> new = pickle.loads(pickle.dumps(approx))
+    >>> new #doctest: +ELLIPSIS
```
Comment (springcoil) [inline pymc3/data.py:124]: I don't understand this - what does this mean?

File: pymc3/tests/test_theanof.py [hunk 2]
```diff
@@ -74,3 +77,23 @@ def gen():
...
+    def test_pickling(self):
+        data = np.random.uniform(size=(1000, 10))
+        minibatches = DataSampler(data, batchsize=50)
+        gen = generator(minibatches)
+        pickle.loads(pickle.dumps(gen))
+        bad_gen = generator(integers())
+        self.assertRaises(Exception, pickle.dumps, bad_gen)
+
+    def test_gen_cloning_with_shape_change(self):
+        data = np.random.uniform(size=(1000, 10))
```
Comment (springcoil) [inline pymc3/tests/test_theanof.py:90]: Wondering (and this is not necessarily a blocker) if some property tests would be interesting here. I think @AustinRochford built some tests like that. 

Just an off the hand the comment. The rest of the tests look fine to me. I've learned a bit from reading your code!

**Review by springcoil — APPROVED** [review]

**Comment by springcoil** [comment]
Seems like the only thing stopping this is coverage/coveralls. Is this an error that's popping up elsewhere? cc @twiecki

File: pymc3/tests/test_theanof.py (hunk 2 above)
Comment (ferrine) [reply pymc3/tests/test_theanof.py:90]: @springcoil this test appeared from a problem that I'm still not able to reproduce. 
After some manipulations with `theano.clone` I have not replaced generator. Thus I was not able to set a new input with new shape (e.g. shared variable)

File: pymc3/data.py (hunk 1 above)
Comment (ferrine) [reply pymc3/data.py:124]: I was doing some tests with Doctest utils, this means that some part of output `...` can change and it is not a problem

**Comment by ferrine** [comment]
As soon as I can reproduce my strange problem with not replaced Generator I'll open an issue/PR reproducing it. Now I thing it's all OK as on most models generator works fine and there is no problem for merge.
BTW We can do minibatch SVGD using these minibatches, currently it has no interface for iterating over minibatches like `Inference` object provides

**Comment by ColCarroll** [comment]
Coveralls appears to have done fine on this PR, not sure why... (#1955)

**Comment by ferrine** [comment]
Wow, that's my case! I found minimal breaking example
```python
def test_gen_double_cloning_with_shape_change():
    dim = 10
    data = np.random.uniform(size=(1000, dim))
    shared_1000 = theano.shared(data)
    constant_50 = tt.constant(data[:50])
    minibatches_50 = DataSampler(data, batchsize=50)
    gen_50 = generator(minibatches_50)
    stream = tt_rng().normal(size=(dim, ))
    stream_in = tt.vector()
    stream_in.tag.test_value = stream.eval()
    const_modif = constant_50.dot(stream_in)
    const_modif_1_50 = theano.clone(const_modif, {stream_in: stream})
    const_modif_2_1000 = theano.clone(const_modif_1_50, {constant_50: shared_1000})
    assert const_modif_1_50.eval().shape == (50,)
    assert const_modif_2_1000.eval().shape == (1000,)
    gen_modif = gen_50.dot(stream_in)
    gen_modif_1_50 = theano.clone(gen_modif, {stream_in: stream})
    gen_modif_2_1000 = theano.clone(gen_modif_1_50, {gen_50: shared_1000})
    assert gen_modif_1_50.eval().shape == (50,)
    assert gen_modif_2_1000.eval().shape == (1000,)
```

**Pull request merged**